nnmnkwii/datasets/jvs.py

```diff
@@ -0,0 +1,221 @@
+            speaker, gender, minf0, maxf0 = fields
+            speaker_info[speaker] = {}
+            speaker_info[speaker]["gender"] = gender
+            speaker_info[speaker]["minf0"] = minf0
+            speaker_info[speaker]["maxf0"] = maxf0
+    return speaker_info
+
+
+class _JVSBaseDataSource(FileDataSource):
+    def __init__(self, data_root, speakers, labelmap, max_files):
+        # only accept "jvs*" format
+        self.data_root = data_root
+        if speakers == "all":
+            speakers = available_speakers
+        for speaker in speakers:
+            if speaker not in available_speakers:
+                raise ValueError("Unknown speaker {}. It should be one of {}".format(speaker, available_speakers))
+        self.speakers = speakers
+        if labelmap is None:
+            labelmap = {s: idx for idx, s in enumerate(speakers)}
+        self.labelmap = labelmap
+        self.labels = None
+        self.max_files = max_files
+        # TODO: currently there are many problem in falset, so we do not support falset
+        self._folders = ["parallel100", "nonpara30", "whisper10"]
+        self._textfilename = "transcripts_utf8.txt"
+        self._wavfoldername = "wav24kHz16bit"
+        self.speaker_info = _parse_speaker_info(data_root)
+        self._nonpara_no_wav = self._validate()
+
+    def _validate(self, strict=False):
+        # for each speaker, validate the data
+        # TODO: check lab file
+        nonpara_no_wav = {}
+        for _, speaker in enumerate(self.speakers):
+            # 1. check folder
+            speaker_folder = join(self.data_root, speaker)
+            folders = listdir(speaker_folder)
+            nonpara_no_wav[speaker] = {}
+            for folder in self._folders:
+                assert folder in folders,\
+                    "Can not find {} for {} in its directory {}"\
+                    .format(folder, speaker, speaker_folder)
+            # 2. check wav and txt file,
+            for folder in self._folders:
+                speaker_textfile = join(speaker_folder, folder, self._textfilename)
+                assert isfile(speaker_textfile), "File {} doesn't exist".format(speaker_textfile)
+                speaker_wavfolder = join(speaker_folder, folder, self._wavfoldername)
+                assert isdir(speaker_wavfolder), "Directory {} doesn't exist".format(speaker_wavfolder)
+                # nonpara
+                if folder == "nonpara30":
+                    with open(speaker_textfile, encoding="utf8") as file_:
+                        txtfiles = [line.strip().split(":")[0] for line in file_]
+                    wavfiles = listdir(speaker_wavfolder)
+                    for file_ in txtfiles:
+                        if file_ + ".wav" not in wavfiles:
+                            nonpara_no_wav[speaker][file_] = '_'
+                    assert (len(txtfiles) - len(nonpara_no_wav[speaker])) == len(wavfiles)
+                # strict mode
+                if strict:
```

Comment:
Could you add explanation of what strict is?

Reply:
check whether each text has a corresponding wavfile, if not, report it.
Some of the wavfiles of text don't exist in this dataset, it doesn't matter for the user, so it is for debugging.